libnmstate/state.py

```diff
@@ -30,6 +30,43 @@
 from libnmstate.prettystate import format_desired_current_state_diff
 from libnmstate.schema import Interface
 from libnmstate.schema import InterfaceType
+from libnmstate.schema import Route
+
+
+class _Route(object):
+    def __init__(self, route):
+        self.state = copy.deepcopy(route)
+        if route.get(Route.STATE) != Route.ABSENT:
+            if Route.TABLE_ID not in route:
+                self.state[Route.TABLE_ID] = Route.USE_DEFAULT_ROUTE_TABLE
+            if Route.METRIC not in route:
+                self.state[Route.METRIC] = Route.USE_DEFAULT_METRIC
+            if Route.NEXT_HOP_ADDRESS not in route:
+                self.state[Route.NEXT_HOP_ADDRESS] = ''
+
+    def __hash__(self):
+        return hash((self.state.get(Route.STATE),
+                     self.state.get(Route.TABLE_ID),
+                     self.state.get(Route.DESTINATION),
+                     self.state.get(Route.NEXT_HOP_INTERFACE),
+                     self.state.get(Route.NEXT_HOP_ADDRESS),
+                     self.state.get(Route.METRIC)))
+
+    def __eq__(self, other):
+        return self.state == other.state
+
+    def is_match(self, other):
+        """
+        Assuming the other state contains all the properties to compare.
```

Comment:
It will be better to avoid assumptions of what is the `other`.
It is ok to define the expectations from `other` in general: self and other states are assumed to contain all properties.
However, if you use attributes instead of a dict, you will not need this at all, the type assures that.
I also do not understand why we need two equality versions (this one and `__eq__`).

Reply:
I cannot use `attributes` as absent route allows missing/wildcard property which attributes/properties cannot handle.

Reply:
How is it different from the current implementation in this regard? A None valued attribute is the same as a missing key in the dict. Do I miss something?
I am not sure how you want to use Route that represent an absent, maybe when that will be clearer other things will get clarified.

Reply:
Treat None as missing property.

Reply:
Function changed to `can_remove()` and comment changed to:
`Return True if other route should be removed by self absent route.`